### PR TITLE
[GH-558]:Fixed github issue "Duplicate notifications in case of mention and assignee" 

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**
**Summary** 

- Duplicate bot DM was getting created when a user is mentioned on a PR/Issue assigned to the user. So created a new notification in that special case for both PR/Issue.

- [Issue #558](https://github.com/mattermost/mattermost-plugin-github/issues/558)

### CODE REVIEW

File: server/plugin/webhook.go [hunk 1]
```diff
@@ -925,7 +925,21 @@ func (p *Plugin) handleCommentMentionNotification(event *github.IssueCommentEven
...
+		assigneeMentioned := false
+		for _, assignee := range assignees {
+			if username == *assignee.Login {
+				assigneeMentioned = true
+				break
+			}
+		}
+
+		if assigneeMentioned {
+			continue
```
Comment (hanzei) [inline server/plugin/webhook.go:940]: Maybe a comment should state where is notification is handled instead.

File: server/plugin/webhook.go [hunk 2]
```diff
@@ -1019,18 +1033,42 @@ func (p *Plugin) handleCommentAssigneeNotification(event *github.IssueCommentEve
...
 	var templateName string
-	switch splitURL[len(splitURL)-2] {
+	switch eventType {
 	case "pull":
 		templateName = "commentAssigneePullRequestNotification"
 	case "issues":
 		templateName = "commentAssigneeIssueNotification"
 	default:
-		p.API.LogWarn("Unhandled issue type", "type", splitURL[len(splitURL)-2])
+		p.API.LogWarn("Unhandled issue type", "Type", eventType)
```
Comment (hanzei) [inline server/plugin/webhook.go:1045]: While we are here...
```suggestion
		p.API.LogDebug("Unhandled issue type", "Type", eventType)
```

File: server/plugin/webhook.go [hunk 3]
```diff
@@ -1019,18 +1033,42 @@ func (p *Plugin) handleCommentAssigneeNotification(event *github.IssueCommentEve
...
+		}
+
+		if usernameMentioned {
+			switch eventType {
+			case "pull":
+				templateName = "commentAssigneeSelfMentionPullRequestNotification"
+			case "issues":
+				templateName = "commentAssigneeSelfMentionIssueNotification"
+			default:
+				p.API.LogWarn("Unhandled issue type", "Type", eventType)
```
Comment (hanzei) [inline server/plugin/webhook.go:1067]:
```suggestion
				p.API.LogDebu("Unhandled issue type", "Type", eventType)
```

**Review by hanzei — APPROVED** [review]
Nice work :+1:

**Review by hanzei — APPROVED** [review]
Nice work :+1:

**Review by hanzei — APPROVED** [review]

File: server/plugin/webhook.go [hunk 4]
```diff
@@ -1019,18 +1034,42 @@ func (p *Plugin) handleCommentAssigneeNotification(event *github.IssueCommentEve
...
+			switch eventType {
+			case "pull":
+				templateName = "commentAssigneeSelfMentionPullRequestNotification"
+			case "issues":
+				templateName = "commentAssigneeSelfMentionIssueNotification"
+			default:
+				p.API.LogDebug("Unhandled issue type", "Type", eventType)
+				return
+			}
+		}
```
Comment (mickmister) [inline server/plugin/webhook.go:1071]: We should make sure we set `templateName` back to its previous value for other iterations of this `assignees` loop.

Also this `default` case will never occur I think

File: server/plugin/template.go [hunk 5]
```diff
@@ -302,6 +302,16 @@ Assignees: {{range $i, $el := .Assignees -}} {{- if $i}}, {{end}}{{template "use
...
 {{.GetComment.GetBody | trimBody | quote | replaceAllGitHubUsernames}}
+`))
+
+	template.Must(masterTemplate.New("commentAssigneeSelfMentionPullRequestNotification").Funcs(funcMap).Parse(`
+{{template "user" .GetSender}} mentioned you on a pull request that you are assigned to {{template "eventRepoIssueFullLinkWithTitle" .}}:
+{{.GetComment.GetBody | trimBody | quote | replaceAllGitHubUsernames}}
+`))
+
+	template.Must(masterTemplate.New("commentAssigneeSelfMentionIssueNotification").Funcs(funcMap).Parse(`
+{{template "user" .GetSender}} mentioned you on a issue that you are assigned to {{template "eventRepoIssueFullLinkWithTitle" .}}:
```
Comment (mickmister) [inline server/plugin/template.go:313]:
```suggestion
{{template "user" .GetSender}} mentioned you on an issue that you are assigned to {{template "eventRepoIssueFullLinkWithTitle" .}}:
```

File: server/plugin/template.go [hunk 6]
```diff
@@ -302,6 +302,16 @@ Assignees: {{range $i, $el := .Assignees -}} {{- if $i}}, {{end}}{{template "use
 	template.Must(masterTemplate.New("commentAssigneeIssueNotification").Funcs(funcMap).Parse(`
 {{template "user" .GetSender}} commented on issue you are assigned to {{template "eventRepoIssueFullLinkWithTitle" .}}:
```
Comment (mickmister) [inline server/plugin/template.go:303]:
```suggestion
{{template "user" .GetSender}} commented on an issue you are assigned to {{template "eventRepoIssueFullLinkWithTitle" .}}:
```

**Review by mickmister — CHANGES_REQUESTED** [review]

File: server/plugin/webhook.go (hunk 4 above)
Comment (Kshitij-Katiyar) [reply server/plugin/webhook.go:1071]: Removed the default and set the templateName

**Comment by Kshitij-Katiyar** [comment]
@hanzei Synced with master

**Review by mickmister — APPROVED** [review]

**Comment by DHaussermann** [comment]
I have a strange situation here where I'm not able to reproduce the issue this PR is solving. 
When both cases are true, I only see the _"Commented on your ..."_ DM  and **not** the _"Mentioned you on..."_   DM. 
I don't know how @mickmister got both. 

Steps:
- Install the marketplace **production release** 
- Create a webhook in target repo
- Connect a test user to GitHub
- Have the test user open a PR or create an issue (I tried both)
- As a 2nd GitHub user - Ensure both _commented on_ and _mention_ events are working for a target issue or PR
- On the target issue or PR add a comment as the 2nd user that includes also a mention for the 1st test user
**Observed**: 1 DM is received about a user commenting

Possibly there is already a solution in place we're not seeing that was included in a release already? 
Unless I've missed a more specific problem I don't understand (@mickmister let me know if there is a more specific point I missed) I don't think we need this PR? Or, is this code cleaner and objectively better than the code that is obsoleted in this change? 
1/5 From my less technical understanding, I'd rather not accept the regression risk if no business value is added. 

cc @hanzei who may also have thoughts.

**Comment by mickmister** [comment]
@DHaussermann This ticket is specifically for the issue/PR _assignee_, as opposed to the author.

**Comment by DHaussermann** [comment]
@mickmister Thank you for pointing out that the bug affects assignee and **not** author. Reproduced on prod build and tested.

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- Assignees of issues and PRs now only get 1 notification DM if there is an issue comment that also includes an at mention.
- Regression tested cases with comment no @ mentions
- Regression testes cases with @ mentions and target user is not author or assigned 

This is solved. No regressions found

LGTM!

**Pull request merged**